internal/pkg/agent/application/upgrade/artifact/download/http/progress_observer.go

```diff
@@ -95,3 +97,36 @@ func (lpObs *loggingProgressObserver) ReportFailed(sourceURI string, timePast ti
 		lpObs.log.Warnf(msg, args...)
 	}
 }
+
+type detailsProgressObserver struct {
+	upgradeDetails *details.Details
+	mu             sync.RWMutex
+}
+
+func newDetailsProgressObserver(upgradeDetails *details.Details) *detailsProgressObserver {
+	upgradeDetails.SetState(details.StateDownloading)
+	return &detailsProgressObserver{
+		upgradeDetails: upgradeDetails,
+	}
+}
+
+func (dpObs *detailsProgressObserver) Report(sourceURI string, timePast time.Duration, downloadedBytes, totalBytes, percentComplete, downloadRate float64) {
+	dpObs.mu.Lock()
+	defer dpObs.mu.Unlock()
+
+	dpObs.upgradeDetails.SetDownloadPercent(percentComplete)
```

Comment:
Why not add all the details? Would it not be helpful to show the speed of the download?

Reply:
I agree that download rate could be useful too.  I will add it and let the Fleet team know to look for that field too so they can show it in the UI.

Reply:
Added in 7d8626746.